### PR TITLE
fix: general repo cleanup

### PR DESCRIPTION
**Pull request by jeff-roche**
- Vendoring
- Containerfile updates to support vendoring and make the file more generic
- Makefile updates to support vendoring
- Removed cruft
    - _OLD directory
    - editor config file

### CODE REVIEW

File: Containerfile [hunk 1]
```diff
@@ -1,19 +1,14 @@
 # SPDX-License-Identifier: Apache-2.0
-
-FROM quay.io/projectquay/golang:1.22 AS makefile
```
Comment (stefanha) [inline Containerfile:3]: This was used by containerized-build.sh, which is being deleted by this PR. Why is it being deleted?

File: Containerfile [hunk 2]
```diff
@@ -28,18 +23,18 @@ COPY deploy/ deploy/
 COPY hack/ hack/
 COPY internal/ internal/
 
-RUN make build
+RUN GOOS=$TARGETOS GOARCH=$TARGETARCH go build -mod=vendor --ldflags "-s -w" -a -o bin/kubesan cmd/main.go
```
Comment (stefanha) [inline Containerfile:26]: Code generation, vetting, and linting no longer runs because the Makefile is no longer being used when building the container image. What is the rationale for this change?

Maybe the idea is that the CI needs to run these things to gate commits that forgot to update generated code or that fail vet/lint? But I don't see that change in this PR so I think we're simply losing these steps.

File: Containerfile [hunk 3]
```diff
@@ -28,18 +23,18 @@ COPY deploy/ deploy/
...
 
 # CentOS Stream 9 doesn't provide package nbd
 # FROM quay.io/centos/centos:stream9
-FROM quay.io/fedora/fedora:40
+FROM --platform=$TARGETPLATFORM fedora:latest
```
Comment (stefanha) [inline Containerfile:30]: Why allow the Fedora version to change externally? This may cause breakage when a new fedora:latest is released. Locking the version number minimizes the risk of build/CI failures caused by external version changes.

File: Containerfile [hunk 4]
```diff
@@ -1,19 +1,14 @@
 # SPDX-License-Identifier: Apache-2.0
-
-FROM quay.io/projectquay/golang:1.22 AS makefile
-
-RUN dnf install -qy diffutils && dnf clean all
+ARG TARGETOS
+ARG TARGETARCH
+ARG TARGETPLATFORM
+FROM golang:1.22 AS builder
```
Comment (stefanha) [inline Containerfile:5]: "--platform=$TARGETPLATFORM " was added to a FROM line below. Is that needed here too?

File: Containerfile (hunk 1 above)
Comment (jeff-roche) [reply Containerfile:3]: I was confused why that file was needed at all, it's just a podman build command. It seems like the same could be achieved by just running a make command if we really want to simplify it:
```
.PHONY: container-build
container-build:
    podman build -t kubesan:latest .
```

File: Containerfile (hunk 4 above)
Comment (jeff-roche) [reply Containerfile:5]: Yes, I'll add it there as well

File: Containerfile (hunk 3 above)
Comment (jeff-roche) [reply Containerfile:30]: No arguments from me, I'll set this back to 40

File: Containerfile (hunk 2 above)
Comment (jeff-roche) [reply Containerfile:26]: Yes that was the idea, I pinged you in slack to discuss CI a little more. I think I need to add some context around what you have for CI

File: Makefile [hunk 5]
```diff
@@ -86,7 +85,7 @@ vet: ## Run go vet.
 
 .PHONY: test-unit
 test-unit: vet lint envtest ## Run unit tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./...
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test -v ./...
```
Comment (stefanha) [inline Makefile:88]: What is the purpose of this change?

File: Containerfile [hunk 6]
```diff
@@ -28,18 +23,18 @@ COPY deploy/ deploy/
...
+RUN dnf update -y && dnf install --nodocs --noplugins -qy nbd qemu-img util-linux-core e2fsprogs xfsprogs && dnf clean all
 
-WORKDIR /kubesan
+WORKDIR /
 
-COPY --from=makefile /kubesan/bin/kubesan bin/
+COPY --from=builder /kubesan/bin/kubesan .
 
-ENTRYPOINT [ "/kubesan/bin/kubesan" ]
+ENTRYPOINT [ "/kubesan" ]
```
Comment (stefanha) [inline Containerfile:40]: I'm fine with this change but can you share the reason for changing the path?

File: Containerfile (hunk 1 above)
Comment (stefanha) [reply Containerfile:3]: containerized-build.sh was added recently because the CI environment does not have the right golang version available. It's the entry point that the Jenkins CI calls.

If things can be simplified by just having the CI call `podman build -t kubesan:latest .` directly, then that would be nice. But I suspect the CI will break if we merge this PR without CI changes.

Adding @fabbione and Chandan Somani to see what they think.

File: Makefile (hunk 5 above)
Comment (jeff-roche) [reply Makefile:88]: To get more output from the unit tests, I know there aren't any currently but that could change in the future and it's nice to get more output (particularly in CI)

File: Containerfile (hunk 6 above)
Comment (jeff-roche) [reply Containerfile:40]: This is more of a nit-picky thing. We usually put the exe in the root but I'm fine leaving it nested

**Comment by ebblake** [comment]
Can we split this commit into smaller pieces?  Addition of the vendor/ directory seems independent from some of the other cleanups.  Likewise, deletion of _OLD/ makes sense, but on its own (so that it is still easy to revert just that one patch when checking if new code needs to borrow any concepts that were working in _OLD before we overhauld the codebase to use kubebuilder but in the process regressed on a lot of features)

File: Containerfile (hunk 1 above)
Comment (ebblake) [reply Containerfile:3]: We definitely want a two-phase build in the Containerfile.  The first half of the file is all about building (pull in whatever dependencies are needed, create scratch files galore, ...); but the second half to create the kubesan image is supposed to be as bare-bones as possible (just the tools and libraries needed to get the driver running, with binaries copied in from the build image).  But we don't necessarily need a full-blown script to call a one-liner, if it is just as easy to write that one-liner into the CI setup directly.

Comment (ebblake) [reply Containerfile:3]: Historical background: before containerized-build.sh was added, 'make build' used to kick off podman unconditionally, but only after also running go to build controller-gen locally.  That mix (the developer's own go building one part of the tools, then the golang in the container running the rest of the build) is not ideal.  containerized-build.sh made it so that we could guarantee that the same version of Go was used for everything in the build.

it would not hurt to have 'make container-build' for developers to use (although the Containerfile must not use it; that would be a recursive nightmare).  Meanwhile, in a CI, you never want to start with bare 'make', but only things that will use the container (but within the container, calling bare 'make' is okay).

File: Containerfile (hunk 2 above)
Comment (csomani1) [reply Containerfile:26]: Since `make build` is no longer part of Containerfile, I am curious how we will call it in CI. I think the CI env should be updated to accommodate these changes before they get merged; otherwise, it will be broken again.

File: Containerfile (hunk 6 above)
Comment (ebblake) [reply Containerfile:40]: I don't care if we un-nest it - as long as it can still be found by the deployment yaml (we are preliminary enough that NOW is the time to sort naming locations, including where important binaries can be found)

File: Containerfile [hunk 7]
```diff
@@ -1,19 +1,14 @@
 # SPDX-License-Identifier: Apache-2.0
-
```
Comment (ebblake) [inline Containerfile:2]: Do we need to name this file Dockerfile? Look at 'git log -- Dockerfile Containerfile' to see a sordid history of different naming approaches we've had over the past (I think we flipped to Dockerfile when gitlab CI wouldn't recognize Containerfile, and know we accidentally flipped back when overhauling to use kubebuilder).  Of course, a rename would be wise as an independent commit.

File: Containerfile [hunk 8]
```diff
@@ -28,18 +23,18 @@ COPY deploy/ deploy/
...
 
 # CentOS Stream 9 doesn't provide package nbd
 # FROM quay.io/centos/centos:stream9
-FROM quay.io/fedora/fedora:40
+FROM --platform=$TARGETPLATFORM fedora:latest
 
 # util-linux-core, e2fsprogs, and xfsprogs are for Filesystem volume support where
 # blkid(8) and mkfs are required by k8s.io/mount-utils.
-RUN dnf install -qy nbd qemu-img util-linux-core e2fsprogs xfsprogs && dnf clean all
+RUN dnf update -y && dnf install --nodocs --noplugins -qy nbd qemu-img util-linux-core e2fsprogs xfsprogs && dnf clean all
```
Comment (ebblake) [inline Containerfile:35]: Nice change. I still have a pending patch that will override nbd-client from this line with a version that implements 'nbd-client -i' (aka netlink /sys/block/nbdN/backend support), but that still works with this cleanup in place.

File: Makefile [hunk 9]
```diff
@@ -55,8 +55,7 @@ help: ## Display this help.
 
 .PHONY: build
 build: .generate.timestamp vet lint ## Build the KubeSAN image.
-	go mod download
-	go build -o bin/kubesan cmd/main.go
+	go build -mod=vendor -o bin/kubesan cmd/main.go
```
Comment (ebblake) [inline Makefile:58]: How does one update the vendor/ directory (such as when intentionally deciding to upgrade to a newer version of a dependency)? Is there a new make target for that?  Meanwhile, the vendor/ directory makes the overall git repo a lot larger, even if it is now safer in that it no longer has to rely on external repositories (or worse supply-chain MitM attackers), so I'm okay with it.  But vendoring should be its own change, independent from the rest of this commit.

File: go.mod [hunk 10]
```diff
@@ -6,8 +6,7 @@ toolchain go1.22.2
 
 require (
 	github.com/container-storage-interface/spec v1.9.0
-	github.com/onsi/ginkgo/v2 v2.17.2
-	github.com/onsi/gomega v1.33.1
+	github.com/digitalocean/go-qemu v0.0.0-20230711162256-2e3d0186973e
 	github.com/openshift/custom-resource-status v1.1.2
```
Comment (ebblake) [inline go.mod:10]: Should go.mod changes be in their own patch (either before or after vendoring)?

File: Makefile (hunk 9 above)
Comment (jeff-roche) [reply Makefile:58]: When I split this out I'll add a make target for it but tl;dr is `go mod tidy && go mod vendor` will do your updates for you

File: vendor/github.com/beorn7/perks/LICENSE [hunk 11]
```diff
@@ -0,0 +1,20 @@
+Copyright (C) 2013 Blake Mizerany
+
+Permission is hereby granted, free of charge, to any person obtaining
+a copy of this software and associated documentation files (the
```
Comment (ebblake) [inline vendor/github.com/beorn7/perks/LICENSE:4]: I'm glad that github hides the vendored files from the diff by default.  I also understand that we don't really have any control over the content of these files (that's the point of vendoring - to copy in a frozen point in time of the dependencies we used).  Do we need to do any automated checking of whether vendored tools have correct Licensing compatible with the rest of our code base?  (Technically, we should already require that everything we depend on was the right license - but now that it is explicitly in our repo, it becomes more obvious if we failed at that by pulling in something with an incompatible license)

Comment (jeff-roche) [reply vendor/github.com/beorn7/perks/LICENSE:4]: We don't usually do this in our repos but I'm not sure if legal has anything to say about it? Something like snyk might check that automatically but I'm not sure off the top of my head

File: go.mod (hunk 10 above)
Comment (jeff-roche) [reply go.mod:10]: I think this can be lumped with vendoring since you would typically run `go mod tidy` with the vendoring command to make sure your deps are all squared away

File: Containerfile (hunk 7 above)
Comment (jeff-roche) [reply Containerfile:2]: It doesn't matter to me either way, for Openshift CI we specify the Dockerfile path so in this case I would just specify `Containerfile`  
example: https://github.com/openshift/release/pull/59746/files#diff-fc2f8ca9e802b37cde660774cd4d43ed044d7d2365ead0ee4f242016f7dcaef2R13

File: Containerfile (hunk 2 above)
Comment (ebblake) [reply Containerfile:26]: The simplest fix to keep CI happy before we go with this full PR:
```
printf 'container-build:\n\tpodman build -t kubesan:latest .\n' >> Makefile
```
and update CI to call 'make container-build' where it now calls './containerized-build.sh'. At that point, containerized-build.sh becomes unused and can be removed again.

Tweak the Makefile line to use proper overrides (for example, we should reinstate the `CONTAINER_TOOL ?= 
podman` line that was lost in commit 37e428, and call $(CONTAINER_TOOL) instead of bare podman)

File: Containerfile (hunk 3 above)
Comment (ebblake) [reply Containerfile:30]: As long as you're looking at this, would you also look at: https://gitlab.com/kubesan/kubesan/-/merge_requests/100/diffs?commit_id=89a370dcde47c8cb2b421245344c147b46984478 and see where I am trying to add 'nbd-client -i' into KubeSAN?  My initial approach was to add a new deps/Containerfile and stash the just-built nbd-client into quay.io/kubesan/nbd-client-i so that not all developers have to repeat the steps - but if we are going to parameterize the build to run multi-arch, then nbd-client-i also needs to be built multi-arch for the same set of architectures.

File: Containerfile (hunk 2 above)
Comment (fabbione) [reply Containerfile:26]: CI call to build is a one liner change. It´s easy enough to me to patch at any given time. Just use the approach you guys prefer and I will just CI.

File: Containerfile (hunk 1 above)
Comment (fabbione) [reply Containerfile:3]: Mentioned in another comment too. CI is here to support you and not to make things more complex. building in CI is a one liner change. Once you decide the approach, I will gladly patch the pipeline.

**Comment by jeff-roche** [comment]
Changes were made upstream
